### DISCUSSION TITLE
What is the recommended way to use both nixpkgs stable and unstable / part-specific nixpkgs extension

### DISCUSSION THREAD

**Discussion by onsails** [hercules-ci/flake-parts#105] [Q&A] [answered]
Before flake-parts, I've ben setting unstable as an overlay
```nix
(self: super: { unstable = import inputs.nixpkgs-unstable { inherit system; }; })
```

I believe the same can be achieved with flake-parts:
```nix
  imports = [
    inputs.flake-parts.flakeModules.easyOverlay
  ];

  perSystem = { config, pkgs, final, system, ... }:
    let
      unstable = import inputs.nixpkgs-unstable {
        inherit system;
      };
    in
    {
      overlayAttrs = {
        inherit unstable;
      };
    };
```

This works well and defined in a special `default` module I import in every nixosConfiguration.

However, in some parts I need to extend packages specificly for those parts, for example set unfreePredicate – which is ignored apparently because easyOverlay forces it's own nixpkgs configuration. Is it possible to

**Comment by roberth** [comment]
> Is it possible to

I think something went wrong with your post.

> However, in some parts I need to extend packages specificly

`easyOverlay` needs to override `pkgs` for the purpose of making `perSystem` behave like an overlay.
Maybe all you really need is to apply an overlay for the flake's own purpose instead of exposing it? In that case a better Nixpkgs module would be sufficient to solve your problem; one with a `nixpkgs.overlays` option that initializes `pkgs` with those, instead of the fancy override business to _publish_ rather than consume an overlay.

**Reply by stites** [reply]
More concretely, I think this suggestion is to package this up in a module like:

```
{config, lib, pkgs, ...}: {
  sound.enable = true;

  environment = {
    etc."xdg/qutebrowser/config.py".text = ...;
    sessionVariables.BROWSER = "qute";
    systemPackages = with pkgs; [qute qutebrowser mpv youtubeDL rofi];
  };
  nixpkgs.overlays = [
    (final: prev: {
      # wrapper to specify config file
      qute = prev.writeShellScriptBin "qute" ''
        QUTE_DARKMODE_VARIANT=qt_515_2 QT_QPA_PLATFORMTHEME= exec ${final.qutebrowser}/bin/qutebrowser -C /etc/xdg/qutebrowser/config.py "$@"
      '';
    })
  ];
}
```

and then import the module, right?

**Comment by stites** [comment, marked as answer]
@onsails, I came across this issue for a different reason than using the `pkgs.unstable` pattern, but I would like to point out that it seems as if the "flake-parts" way is to avoid this pattern of a global namespace as much as possible.

Some more modular ways to do this seem to be:
- @roberth's suggestion (which I think I gave a concrete example of). This does seem to be a little too much effort for me, personally, when I just want a quick, one-off hack for a tool I have not bought into yet.
- maintain a dedicated `pkgs.nix` flake, or use something like https://github.com/fufexan/dotfiles/tree/main/pkgs/default.nix which can use packages via a `self'.packages` property
- Stick to exactly what you are doing, just make sure you place the overlay in the final module you want to work with: https://github.com/fufexan/dotfiles/blob/main/modules/nix.nix#L69-L122

Something I was expecting which did **not** work, which might also have been your assumption, was:
- defining a flake with a `_module.args.pkgs = ...` property and importing it at the top level (as in https://github.com/fufexan/dotfiles/blob/main/flake.nix#L13). This doesn't seem to work for me, but take it with a grain of salt, because I'm also not testing out `easyOverlay` or doing my full due-diligence.
